presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java

```diff
@@ -0,0 +1,419 @@
+                    leftOuterJoin,
+                    groupBySymbols,
+                    aggregations.build(),
+                    functions.build(),
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+
+        public Optional<RewrittenScalarAggregationSource> rewriteScalarAggregationSource(PlanNode aggregationSource, List<Symbol> correlation)
+        {
+            Optional<FilterNode> filter = searchFrom(aggregationSource)
+                    .where(FilterNode.class::isInstance)
+                    .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class))
+                    .findFirst();
+
+            if (!filter.isPresent()) {
+                return Optional.of(new RewrittenScalarAggregationSource(ImmutableList.of(), aggregationSource));
+            }
+
+            Expression predicate = filter.get().getPredicate();
+
+            if (!isSupportedPredicate(predicate)) {
+                return Optional.empty();
+            }
+
+            if (!DependencyExtractor.extractAll(predicate).containsAll(correlation)) {
+                return Optional.empty();
+            }
+
+            ImmutableList.Builder<Expression> newPredicate = ImmutableList.builder();
+            ImmutableList.Builder<Expression> joinExpressions = ImmutableList.builder();
+            for (Expression conjunct : extractConjuncts(predicate)) {
+                List<Symbol> symbols = DependencyExtractor.extractAll(conjunct);
+                if (correlation.stream().anyMatch(symbols::contains)) {
+                    joinExpressions.add(conjunct);
+                }
+                else {
+                    newPredicate.add(conjunct);
+                }
+            }
+
+            FilterNode newFilterNode = new FilterNode(
+                    idAllocator.getNextId(),
+                    filter.get().getSource(),
+                    ExpressionUtils.andOrTrue(newPredicate.build()));
+
+            if (!joinExpressions.build().isEmpty()) {
+                // filter condition has changed so Limit node no longer applies for EXISTS subquery
+                aggregationSource = searchFrom(aggregationSource)
+                        .where(LimitNode.class::isInstance)
+                        .skipOnlyWhen(ProjectNode.class::isInstance)
+                        .removeFirst();
+            }
+            aggregationSource = searchFrom(aggregationSource)
+                    .where(FilterNode.class::isInstance)
+                    .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class))
+                    .replaceAll(newFilterNode);
+
+            aggregationSource = ensureJoinSymbolsAreReturned(aggregationSource, joinExpressions.build());
```

Comment:
Why is this needed? If "aggregationSource" already returns those columns, there's nothing to be done. If it doesn't, we can't synthesize them from thin air (so that would indicate something is seriously broken in the plan structure).